### PR TITLE
feat(copilot): validate custom agents, prompts, hooks, and setup workflow

### LINKED ISSUE #400: enhancement: Copilot custom agents, prompt files, and hooks validation
## Problem / Motivation

GitHub Copilot now supports three new configuration file types that agnix does not detect or validate:

1. **Custom agent definitions** (`.github/agents/*.agent.md`) - Markdown files with YAML frontmatter defining custom Copilot agents. The `description` field is required. Other valid fields include `tools`, `model`, `mcp-servers`, `target`, `argument-hint`, and `handoffs`. Max 30,000 characters.

2. **Reusable prompt files** (`.github/prompts/*.prompt.md`) - Markdown files with optional YAML frontmatter for reusable prompts. Valid frontmatter fields include `description`, `name`, `argument-hint`, `agent`, `model`, and `tools`.

3. **Copilot coding agent hooks** (`.github/hooks/hooks.json`) - JSON configuration for lifecycle hooks on the Copilot coding agent. Has a `version: 1` field and supports 6 event types: `sessionStart`, `sessionEnd`, `userPromptSubmitted`, `preToolUse`, `postToolUse`, `errorOccurred`. Each hook defines `command` (bash/powershell) and `events`.

None of these file types are detected in `crates/agnix-core/src/file_types/detection.rs` or `crates/agnix-core/src/file_types/types.rs`. The `FileType` enum has no variants for them.

## Evidence

- `crates/agnix-core/src/file_types/types.rs`: `FileType` enum has no `CopilotAgent`, `CopilotPrompt`, or `CopilotHooks` variants
- `crates/agnix-core/src/file_types/detection.rs`: No detection patterns for `.github/agents/*.agent.md`, `.github/prompts/*.prompt.md`, or `.github/hooks/hooks.json`
- No existing schemas in `crates/agnix-core/src/schemas/` for these file types
- No existing open issues covering this gap (checked issues #382-385)

### New FileType variants

Add `CopilotAgent`, `CopilotPrompt`, and `CopilotHooks` to the `FileType` enum, with corresponding detection logic.

### New validation rules

**Custom agents (`.github/agents/*.agent.md`)**:
- **COP-007**: Custom agent missing required `description` field in frontmatter
- **COP-008**: Custom agent has unknown/unsupported frontmatter field
- **COP-009**: Custom agent has invalid `target` value (must be `"vscode"` or `"github-copilot"`)
- **COP-010**: Custom agent uses deprecated `infer` field (replaced by user-invokable agents)
- **COP-011**: Custom agent prompt body exceeds 30,000 characters
- **COP-012**: Custom agent uses fields unsupported on GitHub.com (`model`, `argument-hint`, `handoffs`)

**Prompt files (`.github/prompts/*.prompt.md`)**:
- **COP-013**: Prompt file has empty body (no prompt content)
- **COP-014**: Prompt file has unknown/unsupported frontmatter field
- **COP-015**: Prompt file has invalid `agent` mode value

**Hooks (`.github/hooks/hooks.json`)**:
- **COP-017**: Copilot hooks.json validation (schema validation - version field, valid event types, command structure)

**Setup steps**:
- **COP-018**: `copilot-setup-steps.yml` missing required job name (`setup`)

### Implementation outline

1. Add three new `FileType` variants and detection patterns
2. Add schema types in `crates/agnix-core/src/schemas/` for each file type
3. Implement validators in `crates/agnix-core/src/rules/`
4. Add rules COP-007 through COP-018 to `knowledge-base/rules.json` and `knowledge-base/VALIDATION-RULES.md`
5. Register new validators in the registry
6. Add test fixtures under `tests/fixtures/`

## Source

- https://docs.github.com/en/copilot/reference/custom-agents-configuration
- https://docs.github.com/en/copilot/reference/hooks-configuration
- https://code.visualstudio.com/docs/copilot/customization/prompt-files

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary
- add file type detection and enum support for `.github/agents/*.agent.md`, `.github/prompts/*.prompt.md`, `.github/hooks/hooks.json`, and `.github/workflows/copilot-setup-steps.yml`
- implement Copilot validation rules COP-007 through COP-015, COP-017, and COP-018 with new schemas for custom agents, prompt files, hooks, and setup workflow
- fix COP-018 to validate `jobs.copilot-setup-steps` (with Ubuntu `runs-on` support and non-empty `steps`) and sync rule metadata/docs with vendor sources
- expand tests and fixtures for valid and invalid Copilot agent/prompt/hook/workflow files, including parser error branches and edge cases
- update rules parity artifacts and docs (`knowledge-base/rules.json`, `crates/agnix-rules/rules.json`, `knowledge-base/VALIDATION-RULES.md`, `SPEC.md`, `README` family)

## Test Plan
- [x] `cargo fmt`
- [x] `cargo test -p agnix-core --quiet`
- [x] `cargo test -p agnix-cli --test rule_parity --quiet`
- [x] `python scripts/check-rule-counts.py`
- [x] `cargo check --quiet`
- [x] `cargo test --quiet`

## Related Issues
Closes #400

### CODE REVIEW

File: knowledge-base/VALIDATION-RULES.md [hunk 1]
```diff
@@ -1485,8 +1563,8 @@ pub fn validate_skill(path: &Path, content: &str) -> Vec<Diagnostic> {
 
 ---
 
-**Total Coverage**: 157 validation rules across 28 categories
+**Total Coverage**: 168 validation rules across 28 categories
 
 **Knowledge Base**: 11,036 lines, 320KB, 75+ sources
 **Certainty**: 102 HIGH, 52 MEDIUM, 3 LOW
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/VALIDATION-RULES.md]: Fixed in 9db2207 - updated the footer certainty counts to 109 HIGH, 56 MEDIUM, 3 LOW to match the table and current 168-rule totals.

File: knowledge-base/INDEX.md [hunk 2]
```diff
@@ -116,7 +116,7 @@ knowledge-base/
 | Kiro Skills | 1 | 0 | 1 | 0 | 1 |
 | Amp Skills | 1 | 0 | 1 | 0 | 1 |
 | Roo Code Skills | 1 | 0 | 1 | 0 | 1 |
-| **TOTAL** | **157** | **102** | **52** | **3** | **42** |
+| **TOTAL** | **168** | **109** | **56** | **3** | **42** |
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/INDEX.md]: Fixed in 9db2207 - synchronized INDEX total auto-fix count with the detailed rule summary (58).

File: knowledge-base/INDEX.md [hunk 3]
```diff
@@ -95,7 +95,7 @@ knowledge-base/
 | Claude Memory | 12 | 8 | 4 | 0 | 3 |
 | AGENTS.md | 6 | 1 | 5 | 0 | 0 |
 | Claude Plugins | 10 | 8 | 2 | 0 | 2 |
-| GitHub Copilot | 6 | 4 | 2 | 0 | 1 |
+| GitHub Copilot | 17 | 11 | 6 | 0 | 1 |
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/INDEX.md]: Fixed in 9db2207 - corrected the GitHub Copilot auto-fix count to 3 in INDEX to match rules metadata and VALIDATION-RULES summary.

File: crates/agnix-core/src/rules/copilot.rs [hunk 4]
```diff
@@ -49,6 +65,302 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+                    .with_suggestion(format!(
+                        "Remove unknown frontmatter field '{}'.",
+                        unknown.key
+                    )),
+                );
+            }
+        }
+
+        if config.is_rule_enabled("COP-015") {
+            if let Some(schema) = &parsed.schema {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:336]: Fixed in 9db2207 - `validate_reusable_prompt` now emits COP-014 when prompt frontmatter YAML is invalid, and a regression test was added (`test_cop_014_invalid_prompt_frontmatter_yaml`).

File: crates/agnix-core/src/rules/copilot.rs [hunk 5]
```diff
@@ -49,6 +65,316 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+            );
+        }
+    }
+
+    if config.is_rule_enabled("COP-007") {
+        let has_description = parsed
+            .as_ref()
+            .and_then(|p| p.schema.as_ref())
+            .and_then(|s| s.description.as_ref())
+            .is_some_and(|desc| !desc.trim().is_empty());
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:151]: Fixed in 82dcd61 - custom-agent frontmatter parse errors are now reported explicitly as COP-008 invalid YAML (instead of misleading COP-007 missing description), and a regression test was added (`test_cop_008_invalid_agent_frontmatter_yaml`).

File: knowledge-base/INDEX.md [hunk 6]
```diff
@@ -281,7 +281,7 @@ Total Size:           650KB
 Standards Covered:     5 (Agent Skills, MCP, Claude Code, Multi-Platform, Prompt Eng)
 Sources Consulted:    75+ (specs, docs, research papers, repos)
 Research Agents:       5 (10+ sources each)
-Validation Rules:     157 rules
+Validation Rules:     168 rules
 Auto-Fixable Rules:   48 rules
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/INDEX.md]: Addressed in 9285c8f - updated INDEX statistics to \\Auto-Fixable Rules:   58 rules\\.

File: crates/agnix-core/src/rules/copilot.rs [hunk 7]
```diff
@@ -49,6 +65,334 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+                        "Remove 'infer' and use user-invokable custom agents instead.",
+                    ),
+                );
+            }
+
+            if config.is_rule_enabled("COP-012") {
+                let unsupported = [
+                    ("model:", schema.model.is_some(), "model"),
+                    (
+                        "argument-hint:",
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs]: Addressed in 1b61bdc. COP-012 now only applies when the custom agent is targeting GitHub (and is skipped for 	arget: vscode), and I added regression test 	est_cop_012_skips_vscode_target.

File: crates/agnix-core/src/rules/copilot.rs [hunk 8]
```diff
@@ -49,6 +65,335 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+            Diagnostic::error(
+                path.to_path_buf(),
+                line,
+                0,
+                "COP-013",
+                "Prompt file body is empty",
+            )
+            .with_suggestion("Add prompt text below the optional frontmatter."),
+        );
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:311]: Fixed in 9717ca4 - prompt frontmatter parse errors are now handled before COP-013, so invalid YAML no longer produces a misleading empty-body diagnostic.

File: crates/agnix-core/src/rules/copilot.rs [hunk 9]
```diff
@@ -49,6 +65,335 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+                        parsed.start_line,
+                        0,
+                        "COP-008",
+                        format!("Custom agent frontmatter contains invalid YAML: {err}"),
+                    )
+                    .with_suggestion("Fix YAML syntax in custom agent frontmatter."),
+                );
+            }
+            return diagnostics;
+        }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:142]: Addressed in 9717ca4 - invalid custom-agent YAML under COP-008 is now emitted as a warning (matching MEDIUM severity) instead of an error. I kept the existing rule ID to avoid taxonomy churn in this PR, but aligned behavior and messaging for the parse-error path.

File: crates/agnix-core/src/rules/copilot.rs [hunk 10]
```diff
@@ -49,6 +65,335 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+                        parsed.start_line,
+                        0,
+                        "COP-014",
+                        format!("Prompt frontmatter contains invalid YAML: {err}"),
+                    )
+                    .with_suggestion("Fix YAML syntax in prompt frontmatter."),
+                );
+                return diagnostics;
+            }
+
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs]: Addressed in 9717ca4 - prompt frontmatter parse errors under COP-014 are now warnings (not errors), and parse errors short-circuit before COP-013. I kept COP-014 to avoid introducing a new rule ID in this PR while aligning severity and behavior.

File: knowledge-base/VALIDATION-RULES.md [hunk 11]
```diff
@@ -2,7 +2,7 @@
 
 > Consolidated from 320KB knowledge base, 75+ sources, 5 research agents
 
-**Last Updated**: 2026-02-04
+**Last Updated**: 2026-02-13
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply knowledge-base/VALIDATION-RULES.md]: Fixed in 9717ca4 - updated `knowledge-base/VALIDATION-RULES.md` Last Updated to 2026-02-14 for parity with the other updated artifacts.

File: crates/agnix-core/src/schemas/copilot_agent.rs [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+fn find_unknown_keys(yaml: &str, start_line: usize) -> Vec<UnknownKey> {
+    let mut unknown = Vec::new();
+
+    for (i, line) in yaml.lines().enumerate() {
+        let trimmed = line.trim_start();
+        if trimmed.is_empty() || trimmed.starts_with('#') {
+            continue;
+        }
+        if line.starts_with(' ') || line.starts_with('\t') {
+            continue;
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/copilot_agent.rs]: Fixed in 9717ca4 - unknown-key detection now handles uniformly indented frontmatter by detecting top-level indentation first, and I added regression tests for both agent and prompt parsers.

File: crates/agnix-core/src/rules/copilot.rs [hunk 13]
```diff
@@ -49,6 +65,340 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
     }
 }
 
+fn frontmatter_key_line(raw: &str, start_line: usize, key: &str) -> usize {
+    raw.lines()
+        .enumerate()
+        .find(|(_, line)| line.trim_start().starts_with(key))
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs]: Fixed in 0f32a74 - `frontmatter_key_line` now matches key names with optional whitespace before the colon (e.g. `target : ...`), so COP-009/010/012/015 report the correct line.

File: crates/agnix-core/src/rules/copilot.rs [hunk 14]
```diff
@@ -49,6 +65,340 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+            Diagnostic::error(
+                path.to_path_buf(),
+                line,
+                0,
+                "COP-013",
+                "Prompt file body is empty",
+            )
+            .with_suggestion("Add prompt text below the optional frontmatter."),
+        );
+    }
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:311]: Fixed in 0f32a74 - COP-013 now clamps its reported line to the file length to avoid out-of-range positions when the file ends right after frontmatter.

File: crates/agnix-core/src/rules/copilot.rs [hunk 15]
```diff
@@ -49,6 +65,340 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+        if !has_description {
+            diagnostics.push(
+                Diagnostic::error(
+                    path.to_path_buf(),
+                    1,
+                    0,
+                    "COP-007",
+                    "Custom agent frontmatter is missing required 'description' field",
+                )
+                .with_suggestion("Add a non-empty 'description' key in YAML frontmatter."),
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs]: Fixed in 0f32a74 - COP-007 now emits a dedicated message/suggestion when frontmatter is missing entirely, instead of the generic missing-description wording.

File: crates/agnix-core/src/schemas/copilot_agent.rs [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+    }
+
+    let lines: Vec<&str> = content.lines().collect();
+    if lines.is_empty() {
+        return None;
+    }
+
+    let mut end_idx = None;
+    for (i, line) in lines.iter().enumerate().skip(1) {
+        if line.trim() == "---" {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/copilot_agent.rs:89]: Fixed in 0f32a74 - frontmatter closing fence detection in the agent parser now requires the fence at column 0 (`---`), so indented `---` inside YAML block scalars no longer terminates frontmatter early.

File: crates/agnix-core/src/rules/copilot.rs [hunk 17]
```diff
@@ -49,6 +65,340 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+            );
+        }
+    }
+
+    if let Some(parsed) = &parsed {
+        if let Some(err) = &parsed.parse_error {
+            if config.is_rule_enabled("COP-008") {
+                diagnostics.push(
+                    Diagnostic::warning(
+                        path.to_path_buf(),
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:121]: Fixed in 0f32a74 - when COP-008 is disabled, agent frontmatter parse errors now fall back to COP-007 so invalid YAML is still surfaced instead of silently accepted.

File: crates/agnix-core/src/schemas/copilot_agent.rs [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+    }
+
+    let lines: Vec<&str> = content.lines().collect();
+    if lines.is_empty() {
+        return None;
+    }
+
+    let mut end_idx = None;
+    for (i, line) in lines.iter().enumerate().skip(1) {
+        if *line == "---" {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/copilot_agent.rs]: Fixed in 820bd38 - agent frontmatter closing-fence detection now accepts whitespace around --- while still avoiding false closes on indented block-scalar content via indentation-aware matching.

File: crates/agnix-core/src/schemas/copilot_prompt.rs [hunk 19]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    let lines: Vec<&str> = content.lines().collect();
+    if lines.is_empty() {
+        return None;
+    }
+
+    let mut end_idx = None;
+    for (i, line) in lines.iter().enumerate().skip(1) {
+        if *line == "---" {
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/schemas/copilot_prompt.rs]: Fixed in 820bd38 - prompt frontmatter now uses the same indentation-aware whitespace-tolerant closing-fence detection, so  ---/---  are accepted without false unterminated-YAML diagnostics.

File: crates/agnix-core/src/rules/copilot.rs [hunk 20]
```diff
@@ -202,29 +589,25 @@ impl Validator for CopilotValidator {
...
+                                0,
+                                "COP-003",
+                                t!(
+                                    "rules.cop_003.message",
+                                    pattern = apply_to.as_str(),
+                                    error = validation.error.unwrap_or_default()
+                                ),
+                            )
+                            .with_suggestion(t!("rules.cop_003.suggestion")),
+                        );
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:610]: Addressed in 0fe1e9a - updated `tests/fixtures/copilot/.github/instructions/frontend/react.instructions.md` to use a valid single glob (`**/*.tsx`) so it aligns with current COP-003 validation semantics.

File: crates/agnix-mcp/tests/mcp_tests.rs [hunk 21]
```diff
@@ -233,8 +233,8 @@ mod rules_tests {
 
     #[test]
     fn test_rules_count() {
-        // Should have 170 rules
-        assert_eq!(agnix_rules::rule_count(), 170);
+        // Should have 181 rules
+        assert_eq!(agnix_rules::rule_count(), 181);
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-mcp/tests/mcp_tests.rs:237]: Addressed in 4e12949 - updated `crates/agnix-mcp/tests/mcp_tests.rs` to assert `agnix_rules::rule_count() == 181`.

File: crates/agnix-core/src/rules/copilot.rs [hunk 22]
```diff
@@ -52,6 +65,368 @@ fn line_byte_range(content: &str, line_number: usize) -> Option<(usize, usize)>
...
+        }
+    }
+
+    if let Some(parsed) = &parsed {
+        if let Some(err) = &parsed.parse_error {
+            if config.is_rule_enabled("COP-008") {
+                diagnostics.push(
+                    Diagnostic::warning(
+                        path.to_path_buf(),
+                        parsed.start_line,
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply crates/agnix-core/src/rules/copilot.rs:122]: Addressed in 5c0a3f7 - custom-agent frontmatter parse failures now emit `COP-008` as an error-level diagnostic, and the test `test_cop_008_invalid_agent_frontmatter_yaml` now asserts `DiagnosticLevel::Error`.
